tests/PHPStan/Analyser/data/bug-7621-2.php

```diff
@@ -0,0 +1,23 @@
+<?php declare(strict_types = 1);
+
+namespace Bug7621;
+
+use function PHPStan\Testing\assertType;
+
+class Foo
+{
+	private const FOO = [ 'foo' => ['foo', 'bar'] ];
+
+	public function foo(): void
+	{
+		assertType('array{\'foo\', \'bar\'}', self::FOO['foo']);
+		$keys = [0, 1, 2];
+		foreach ($keys as $key) {
+			if (array_key_exists($key, self::FOO['foo'])) {
+				assertType('array{\'foo\', \'bar\'}', self::FOO['foo']);
+			} else {
+				assertType('array{0?: \'foo\', 1?: \'bar\'}&non-empty-array', self::FOO['foo']);
```

Comment:
that's a weird intersection. apparently `{1: bar}|{0: foo}|{0: foo, 1: bar}` leads to this. I could verify that `union` adds this. maybe a sorting problem hmm

Reply:
or it actually should be like that. was added in https://github.com/phpstan/phpstan-src/commit/036056bb#diff-fef252fe8071200a31b3d282d1442135d62297098c9b20fd731f3888398ee973 and trying to "optimise" it away breaks things

Reply:
irrelevant, has been fixed by Ondrej in the tagged unions preparations and tests were added in https://github.com/phpstan/phpstan-src/pull/1553